Another stab at groupWithin

changlinli opened the pull request:
This take tries to add several new features:

+ Better handling of segments, not breaking them up when not necessary
+ More accurate buffering: we only ever buffer up to the specified n
  elements instead of potentially 2n elements
+ No integer overflow to deal with by using arbitrary java Objects for
  comparison instead of Ints

SystemFw committed:
`groupWithin` prototype

SystemFw committed:
`startTimeout` is required at the beginning of the Stream

SystemFw committed:
Handle interruption of the source stream

SystemFw committed:
Add tests

SystemFw committed:
Fix off-by-1 error

changlinli committed:
Another stab at groupWithin

This take tries to add several new features:

+ Better handling of segments, not breaking them up when not necessary
+ More accurate buffering: we only ever buffer up to the specified n
  elements instead of potentially 2n elements
+ No integer overflow to deal with by using arbitrary java Objects for
  comparison instead of Ints

Daenyth commented on core/shared/src/main/scala/fs2/Stream.scala:
```diff
@@ -1902,52 +1923,98 @@ object Stream {
-      * Note: a time window starts each time downstream pulls.
+      * Empty groups can occur if no elements can be pulled from upstream
+      * in a given time window. If you would like to have only non-empty groups,
+      * you can simply `collect` out the empty groups (see the provided example).
```
Minor style nit; `simply` is a useless word - if what you explain is simple, then the explanation isn't helped by having the word. If it's not actually simple to the reader then it comes off condescending.

changlinli replied:
Very true. One second.

Daenyth reviewed (approved):
The implementation here is nice and easy to follow 👍

changlinli committed:
Further improve readability

Name the actions we are trying to do and include a comment for why we
continue accumulating when we see an unrecognized timeout.

changlinli committed:
Fix early recursion in groupWithinV

Turns out we need to manually trampoline if we want to extract out the
recursive call.

SystemFw commented:
@changlinli I've having a hard time understanding how this new implementation differs from the existing one wrt 

> More accurate buffering: we only ever buffer up to the specified n elements instead of potentially 2n elements

Could you expand a bit please?

Where I'm coming from is trying to understand if the additional complexity is justified or not (might well be, I don't know yet)

changlinli commented:
Previously the accumulator plus the queue size meant that you would have `2n` elements buffered (as well as potentially a latency of `2 * duration` between when the element is pulled and is passed down downstream). In this case there is no queue of elements (only a singleton queue that stores the timeout), only the accumulator. Instead we read directly from the stream itself and merge with the singleton queue.

Also most of the code difference is in dealing with `unconsLimit` and my more verbose variable names/pulling out intermediate recursive calls. I actually think most of the complexity goes away if you don't use `unconsLimit`.

SystemFw commented:
> In this case there is no queue of elements (only a singleton queue that stores the timeout), only the accumulator. Instead we read directly from the stream itself and merge with the singleton queue.

See, but this isn't true. There is a queue, in `merge` (internally). A segment gets pulled in the the merge queue (which is a queue of segments), and stays there until it's pulled. When you call `unconsLimit`, it doesn't just get `n` elems max (impossible to do without breaking the segment),  it calls `uncons` (thus getting a whole segment), and then splits it. But by calling `uncons`, it frees a spot in the queue of segments in `merge`, which then gets another segment, so you're still buffering more than you think.

> I actually think most of the complexity goes away if you don't use unconsLimit.

How does this code work without `unconsLimit`? Seems crucial for this implementation

changlinli commented:
Yep, it's certainly true that no matter what implementation you have, a pull-based approach means you'll always have at least one segment buffered (because you could be in the middle of pulling and get beat by your timer). So you're right it's usually more than `n` (but hopefully the minimum over `n` that you could get because of that issue).

As for `unconsLimit`, you could just do `uncons1` and then get rid of the nasty `fold` I do (which really is just getting around the fact that I have a `Segment[Either[TimeoutTick, O], Unit]` when because of the semantics of the `merge` I really have a `Either[Segment[O, Unit], Segment[TimeoutTick, Unit]]` and the right-hand side is a singleton.

SystemFw commented:
@mpilquist mind giving a second opinion on this? I'd like to get it over the finish line

mpilquist commented on core/shared/src/main/scala/fs2/Stream.scala:
```diff
@@ -1730,6 +1730,27 @@ object Stream {
+    /**
+      * Like `either`, but halts as soon as `self` halts.
+      */
+    def eitherHaltL[O2](that: Stream[F, O2])(implicit F: Effect[F],
```
We'll need this in the PureOps class too.

changlinli replied:
Do you think this should exist independently of what happens with `groupWithin` (i.e. if I somehow am able to remove the dependence on `Either`)? If so I'm happy to add it there.

mpilquist commented on core/shared/src/main/scala/fs2/Stream.scala:
```diff
@@ -1902,52 +1923,105 @@ object Stream {
-      }
+      for {
+        tickQueue <- Stream.eval(async.mutable.Queue.synchronous[F, TimeoutTick](F, ec))
+        mergedStreams = tickQueue.dequeue.eitherHaltR(self)
```
Is there a way we could do this without using Either or otherwise transforming the primary stream segments? Consider a stream consisting of huge byte array backed chunks. Mapping each byte through `Right(_)` is going to be a huge performance cost.

changlinli replied:
That's fair. I'm personally not as worried about a performance hit because you're already doing a time-based wait so presumably this isn't a particular performance-sensitive place, but it makes sense not to mindlessly throw away performance. I'll think about it some more. The only solution I can think of off the top of my head is to use `null` as a tick and check for that, but that feels a bit icky.

mpilquist commented on core/shared/src/main/scala/fs2/Stream.scala:
```diff
@@ -1902,52 +1923,105 @@ object Stream {
+            case None => Pull.output1(acc)
+            case Some((segment, restOfStream)) =>
+              val (timeoutOpt, elems) =
+                segment.force.toVector.foldLeft((Option.empty[TimeoutTick], Vector.empty[O])) {
```
I'd rather see the main implementation be Segment based and then derive a Vector version from that one. I suspect that will have much better performance.

changlinli replied:
I'm still not sure I understand exactly how the performance of `Segment` works out. If I build up a bunch of `Catenable` `Segment`s with `++`, how does the chunk structure work out? Do all of the underlying chunks behind the `Segment`s get appended together?

mpilquist reviewed (commented):
Added some comments. General approach looks good to me. I'm a bit concerned about boxing/unboxing and unchunking in this implementation, especially when coupled with a large primitive chunks.

changlinli commented:
@SystemFw @mpilquist Apologies on being away for a bit. Just saw the notification in the main fs2 repo. What's needed to push this over the line? Moving it to a `Segment`-based implementation? Should have some time over the weekend to do whatever needs to be done.

SystemFw commented:
Closing this as I went for a different implementation in https://github.com/functional-streams-for-scala/fs2/pull/1158
